### ISSUE TITLE
Custom anchor boxes and mask index

### ISSUE THREAD

**Issue by nasio91** [AlexeyAB/darknet#5887]
Hi i'm training on custom dataset and i calcultaed my own anchor boxes on this. I got this set of values: 
11, 18,  18, 30,  26, 45,  40, 64,  56, 96,  84,120,  92,178, 136,223, 200,296 

The default cfg file make me a little bit confused about the mask values: 

in the first Yolo layer i see that 
[yolo]
mask = 6,7,8
anchors = 10,13,  16,30,  33,23,  30,61,  62,45,  59,119,  116,90,  156,198,  373,326
classes=3
num=9
jitter=.3
ignore_thresh = .7
truth_thresh = 1
random=1

In the "how to train" i read "1st-[yolo]-layer has anchors larger than 60x60, 2nd larger than 30x30, 3rd remaining"  but the fifth value 59,119 isn't greate than 60*60? 

So following exactly what is written leaving out the cfg file. I have to set: 
1 layer mask=4,5,6,7,8
2 layer mask =2,3 
3 layer mask=0,1 

It's correct?

**Comment by AlexeyAB** [comment]
Preferably to compare 33x23=759 and 30x30=900, so 33x23 < 30x30
And 59x119 > 60x60

So for yolov3:
1 layer mask= 5,6,7,8
2 layer mask =3,4
3 layer mask= 0,1,2

For yolov4 - vice versa:
1 layer mask= 0,1,2
2 layer mask =3,4
3 layer mask= 5,6,7,8

**Comment by ramdhan1989** [comment]
> Preferably to compare 33x23=759 and 30x30=900, so 33x23 < 30x30
> And 59x119 > 60x60
> 
> So for yolov3:
> 1 layer mask= 5,6,7,8
> 2 layer mask =3,4
> 3 layer mask= 0,1,2
> 
> For yolov4 - vice versa:
> 1 layer mask= 0,1,2
> 2 layer mask =3,4
> 3 layer mask= 5,6,7,8

Just for clarification, so what stated in "how to train" 

>  But you should change indexes of anchors masks= for each [yolo]-layer, so that 1st-[yolo]-layer has anchors larger than 60x60, 2nd larger than 30x30, 3rd remaining.

is it applicable for yolov3 only ? then, what is the rule for anchors masks yolov4 ? my understanding from your answer above to get anchors masks yolov4 is just exchanging the index layer 1 with layer 3 and vice versa from yolov3 cfg? is my understanding correct ?

thank you

**Comment by helen12138** [comment]
@AlexeyAB  Excuse me,  I got anchors with command 
./darknet detector calc_anchors data/obj.data -num_of_clusters 9 -width 608 -height 864 -show
The anchors:  93, 31, 116, 81, 235, 42, 300, 89, 170,168, 338,198, 211,451, 385,312, 397,606
There is no  size of anchor smaller than 30*30,  so how to set mask index in 1st yolo layer?

**Comment by AlexeyAB** [comment]
Yes, this is normal, just add the first anchor 30x30, and set mask=0 for the first layer.

https://github.com/AlexeyAB/darknet#how-to-improve-object-detection

Only if you are an expert in neural detection networks - recalculate anchors for your dataset for width and height from cfg-file: darknet.exe detector calc_anchors data/obj.data -num_of_clusters 9 -width 416 -height 416 then set the same 9 anchors in each of 3 [yolo]-layers in your cfg-file. But you should change indexes of anchors masks= for each [yolo]-layer, so for YOLOv4 the 1st-[yolo]-layer has anchors smaller than 30x30, 2nd smaller than 60x60, 3rd remaining, and vice versa for YOLOv3. Also you should change the filters=(classes + 5)*<number of mask> before each [yolo]-layer. If many of the calculated anchors do not fit under the appropriate layers - then just try using all the default anchors.

**Comment by annalisasheehan** [comment]
Hi, I was wondering if there is a minimum number of anchors? I'm using yolo to detect vehicles in satellite imagery and therefore the objects are quite small and similar sizes. I ran k-means on my training dataset which suggested there are 3 clusters in my dataset. Can I use 3 anchors for YOLO? Can I set the mask as the 3 anchors in all 3 YOLO layers?
